### PR TITLE
Consistent API for field locations

### LINKED ISSUE #659: API divergence for abstract operations versus boundary conditions
Our API is currently inconsistent. The design for AbstractOperations assumes that field locations are specified with *uninstantiated* types:

```julia
julia> using Oceananigans, Oceananigans.Fields, Oceananigans.AbstractOperations

julia> arch = CPU(); grid = RegularCartesianGrid(size=(16, 16, 16), length=(1, 1, 1));

julia> c, d = (CellField(arch, grid) for i=1:2);

julia> e = @at (Face, Cell, Cell) c * d
BinaryOperation at (Face, Cell, Cell)
├── grid: RegularCartesianGrid{Float64,Periodic,Periodic,Bounded,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}
│   ├── size: (16, 16, 16)
│   └── domain: x ∈ [0.0, 1.0], y ∈ [0.0, 1.0], z ∈ [-1.0, 0.0]
└── tree: 

* at (Face, Cell, Cell) via ℑxᶠᵃᵃ
├── OffsetArray{Float64, 3, Array{Float64,3}}
└── OffsetArray{Float64, 3, Array{Float64,3}}
```

notice that the "location" is specified with `(Face, Cell, Cell)`.

Boundary conditions are different:

```julia
julia> bc = FieldBoundaryConditions(grid, (Cell, Cell, Cell), north=BoundaryCondition(Value, 1.0))
ERROR: MethodError: no method matching DefaultBoundaryCondition(::Bounded, ::Type{Cell})
Closest candidates are:
  DefaultBoundaryCondition(::Bounded, ::Face) at /Users/gregorywagner/.julia/packages/Oceananigans/Pyzcd/src/BoundaryConditions/field_boundary_conditions.jl:20
  DefaultBoundaryCondition(::Bounded, ::Cell) at /Users/gregorywagner/.julia/packages/Oceananigans/Pyzcd/src/BoundaryConditions/field_boundary_conditions.jl:19
  DefaultBoundaryCondition(::Union{Flat, Periodic}, ::Any) at /Users/gregorywagner/.julia/packages/Oceananigans/Pyzcd/src/BoundaryConditions/field_boundary_conditions.jl:17
Stacktrace:
 [1] (::Core.var"#kw#Type")(::NamedTuple{(:north,),Tuple{BoundaryCondition{Value,Float64}}}, ::Type{NamedTuple{(:x, :y, :z),T} where T<:Tuple}, ::RegularCartesianGrid{Float64,Periodic,Periodic,Bounded,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}, ::Tuple{DataType,DataType,DataType}) at ./none:0
 [2] top-level scope at REPL[13]:1
```

while this works:

```julia
julia> bc = FieldBoundaryConditions(grid, (Cell(), Cell(), Cell()), north=BoundaryCondition(Value, 1.0))
Oceananigans.FieldBoundaryConditions (NamedTuple{(:x, :y, :z)}), with boundary conditions
├── x: CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing},BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing}}
├── y: CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic,Nothing},BoundaryCondition{Value,Float64}}
└── z: CoordinateBoundaryConditions{BoundaryCondition{Flux,Nothing},BoundaryCondition{Flux,Nothing}}
```

where the location is specified with `(Cell(), Cell(), Cell())`. 

I'm fine with either API choice (instantiated versus uninstantiated, but we should pick one and stick with it.

I chose uninstantiated types for `AbstractOperations` because I thought the user API was less cluttered that way (fewer parentheses). But there are internal advantages to requiring instantiated types (and this is how most julia software works).

### PR DESCRIPTION
**Pull request by glwagner**
This PR refactors the FieldBoundaryConditions constructor to allow a consistent user API for specifying field locations. Field locations are specified with _uninstantiated_ types: in other words, a location is 

```julia
(Face, Cell, Cell)
```

rather than

```julia
(Face(), Cell(), Cell())
```

The latter form was deemed too much `()`. This is open to debate of course. I am slightly worried the former form is non-Julian. But it seems fairly unequivocal that it's easier to read.

I also took the opportunity to clean up field constructors a bit. We now have the concept of _default_ boundary conditions; mostly, this is useful when a user wants to build a field that will never call `fill_halo_regions` and therefore its boundary conditions are irrelevant. I changed the `two_dimensionsional_turbulence.jl` example to use this default.

The constructor shenanigans aren't perfect, and there is still a constructor that actually does accept instantiated locations (despite that we officially discourage it, and instantiated field locations are not supported for boundary conditions or abstract operations). It seems the checkpointer needs this special constructor.

Resolves #659  .

### CODE REVIEW

File: examples/two_dimensional_turbulence.jl [hunk 1]
```diff
@@ -59,7 +59,7 @@ nothing # hide
 # `vorticity_operation` we create a field `ω` to store the result of the operation, and a
 # `Computation` object for coordinate the computation of vorticity and storage in `ω`:
 
-ω = Field(Face, Face, Cell, model.architecture, model.grid, TracerBoundaryConditions(model.grid))
+ω = Field(Face, Face, Cell, model.architecture, model.grid)
```
Comment (ali-ramadhan) [inline examples/two_dimensional_turbulence.jl]: This change will be helpful

File: src/Fields/field.jl [hunk 2]
```diff
@@ -49,101 +49,110 @@ struct Field{X, Y, Z, A, G, B} <: AbstractField{X, Y, Z, A, G}
...
+                                 data = zeros(arch, grid, (X, Y, Z)) ] )
+
+Construct a `Field` on `grid` with `data` on architecture `arch` with
+boundary conditions `bcs`. Each of `(X, Y, Z)` is either `Cell` or `Face` and determines 
+the field's location in `(x, y, z)`.
+
+Example
+=======
+
+julia> ω = Field(Face, Face, Cell, CPU(), RegularCartesianmodel.grid)
```
Comment (ali-ramadhan) [inline src/Fields/field.jl]:
```suggestion
julia> ω = Field(Face, Face, Cell, CPU(), model.grid)
```

**Comment by ali-ramadhan** [comment]
> We now have the concept of _default_ boundary conditions; mostly, this is useful when a user wants to build a field that will never call `fill_halo_regions` and therefore its boundary conditions are irrelevant.

That makes a lot of sense for a default.

> The constructor shenanigans aren't perfect, and there is still a constructor that actually does accept instantiated locations (despite that we officially discourage it, and instantiated field locations are not supported for boundary conditions or abstract operations). It seems the checkpointer needs this special constructor.

Hmmm, I had a quick look at the checkpointer but couldn't see any instances of `Cell` or `Face`, but we can clean it up in a later PR.

**Comment by glwagner** [comment]
> Hmmm, I had a quick look at the checkpointer but couldn't see any instances of Cell or Face, but we can clean it up in a later PR.

I think its here:

https://github.com/climate-machine/Oceananigans.jl/blob/229b35d3c49144b518c192105e5e7ea94247d7e6/src/OutputWriters/checkpointer.jl#L107

There is also a typo: `field_address` should be `field_location`. Anyways, is `field_address` a tuple?

**Comment by ali-ramadhan** [comment]
Yeah it should be `field_location`.

Ah it's a tuple of instantiated types:
https://github.com/climate-machine/Oceananigans.jl/blob/229b35d3c49144b518c192105e5e7ea94247d7e6/src/OutputWriters/output_writer_utils.jl#L51-L55

So we should change that to
```julia
serializeproperty!(file, location * "/location", (LX, LY, LZ)) 
```

**Comment by glwagner** [comment]
Ah. Maybe we can avoid instantiation then.

**Pull request merged**